src/directives/badge/badge.directive.ts

```diff
@@ -0,0 +1,214 @@
+    },
+})
+export class BadgeDirective implements AfterViewInit, OnChanges, OnDestroy {
+    private readonly _className = 'ux-badge';
+    private readonly _darkColor: ThemeColor = ThemeColor.parse('#000');
+    private readonly _lightColor: ThemeColor = ThemeColor.parse('#FFF');
+    private _badgeElement: HTMLElement | undefined;
+    private _isNumber: boolean;
+
+    /**
+     * Directive parameter that sets the content of the badge
+     */
+    private _badgeDisplayContent: string;
+
+    @Input('uxBadge')
+    get badgeContent(): string {
+        return this._badgeContent;
+    }
+    set badgeContent(s: string) {
+        if (s && s.replace(/ /g, '').length > 0) {
+            const subject = s.toString().trim();
+            this._isNumber = /^\d+$/.test(subject);
+            this._badgeContent = subject;
+        } else {
+            this._badgeContent = null;
+        }
+    }
+    private _badgeContent: string = null;
+
+    /**
+     * Define the badge background color
+     */
+    @Input()
+    get badgeColor(): string {
+        return this._badgeColor.toHex();
+    }
+    set badgeColor(color: string) {
+        this._badgeColor = this.parseThemeColor(color);
+    }
+    private _badgeColor: ThemeColor = this._darkColor;
+
+    /**
+     * Define the badge border color - if unset there is no border
+     */
+    @Input()
+    get badgeBorderColor(): string {
+        return this._badgeBorderColor.toHex();
+    }
+    set badgeBorderColor(color: string) {
+        this._badgeBorderColor = this.parseThemeColor(color);
+    }
+    private _badgeBorderColor: ThemeColor;
+
+    /**
+     * Set the badge vertical position in relation to the parent element
+     */
+    @Input()
+    badgeVerticalPosition: BadgeVerticalPosition = 'above';
+
+    /**
+     * Set the badge horizontal position in relation to the parent element
+     */
+    @Input()
+    badgeHorizontalPosition: BadgeHorizontalPosition = 'after';
+
+    /**
+     * Set if the badge overlaps parent content or flows after parent
+     */
+    @HostBinding('class.ux-badge-overlap')
+    @Input()
+    badgeOverlap: boolean = false;
+
+    /**
+     * Max value that can be displayed - if string measures no. of characters,
+     * if number checks against exact number not no. of characters
+     */
+    @Input() badgeMaxValue: number;
+
+    /**
+     * Badge size (based on CSS styles)
+     */
+    @Input()
+    badgeSize: BadgeSize = 'medium';
+
+    /**
+     * Hide badge from view
+     */
+    @HostBinding('class.ux-badge-hidden')
+    @Input()
+    badgeHidden: boolean = false;
+
+    constructor(
+        private readonly _element: ElementRef<HTMLElement>,
+        private readonly _renderer: Renderer2,
+        private readonly _colorService: ColorService,
+        private readonly _contrastService: ContrastService
+    ) { }
+
+    ngAfterViewInit(): void {
+        const badgeElement: HTMLSpanElement = this._renderer.createElement('span');
```

Comment:
The lifecycle hook functions are getting a bit long and have a bit of duplicated code. Perhaps we could split them up a bit, to make them easier to maintain eg:

```typescript
private createBadge(): void {}
private setBadgeColor(): void {}
private setBadgeBorderColor(): void {}
...
```

That way we can move all the logic for each into its one function, each with a single responsibility.